tls.go

```diff
@@ -112,6 +121,45 @@ func (c *certificate) reload() error {
 	return nil
 }
 
+func (c *certificate) reloadFromPKCS11() error {
+	// Expecting keystore file to only have certificate,
+	// with the private key being in an HSM/PKCS11 module.
+	keystore, err := os.Open(c.keystorePath)
+	if err != nil {
+		return err
+	}
+
+	certAndKey := tls.Certificate{}
+	err = certigo.ReadAsX509FromFiles(
+		[]*os.File{keystore}, "", nil,
+		func(cert *x509.Certificate) {
+			if certAndKey.Leaf == nil {
+				certAndKey.Leaf = cert
+			}
+			certAndKey.Certificate = append(certAndKey.Certificate, cert.Raw)
+		})
+	if err != nil {
+		return err
+	}
+
+	// Reuse previously loaded PKCS11 private key if we already have it. We want to
+	// avoid reloading the key every time the cert reloads, as it's a potentially
+	// expensive operation that calls out into a shared library.
+	if c.cached != nil {
+		old, _ := c.getCertificate(nil)
```

Comment:
That makes sense; if we ever change `getCertificate`'s implementation so that it returns an meaningful error, though, it could be confusing to debug when the error is ignored.